fail to link op/avx component with nvhpc compilers

As reported by Ray Muno on the user mailing list (https://www.mail-archive.com/users@lists.open-mpi.org/msg34594.html), op/avx component cannot be linked when the nvhpc compilers are used

```
 CCLD     liblocal_ops_avx512.la
   CCLD     mca_op_avx.la
./.libs/liblocal_ops_avx512.a(liblocal_ops_avx512_la-op_avx_functions.o):(.data+0x0): multiple 
definition of `ompi_op_avx_functions_avx2'
./.libs/liblocal_ops_avx2.a(liblocal_ops_avx2_la-op_avx_functions.o):(.data+0x0): first defined here
./.libs/liblocal_ops_avx512.a(liblocal_ops_avx512_la-op_avx_functions.o): In function 
`ompi_op_avx_2buff_min_uint16_t_avx2':
/project/muno/OpenMPI/BUILD/SRC/openmpi-4.1.1/ompi/mca/op/avx/op_avx_functions.c:651: multiple 
definition of `ompi_op_avx_3buff_functions_avx2'
./.libs/liblocal_ops_avx2.a(liblocal_ops_avx2_la-op_avx_functions.o):/project/muno/OpenMPI/BUILD/SRC/openmpi-4.1.1/ompi/mca/op/avx/op_avx_functions.c:651: 
first defined here
make[2]: *** [mca_op_avx.la] Error 2
```

The root cause is Open MPI assumes the following macros are defined if the compiler supports the AVX512 features we expect
```
__AVX512BW__
 __AVX512F__
__AVX512VL__
```
and at least one of these macros is not defined by nvhpc compilers


From `ompi/mca/op/avx/op_avx_functions.c` 

```c
#if defined(GENERATE_AVX512_CODE)
#  if defined(__AVX512BW__) && defined(__AVX512F__) && defined(__AVX512VL__)
#    define PREPEND _avx512
#  else
#    undef GENERATE_AVX512_CODE
#  endif  /* defined(__AVX512BW__) && defined(__AVX512F__) && defined(__AVX512VL__) */
#endif  /* defined(GENERATE_AVX512_CODE) */
```

I am wondering whether we really need to re-test these macros. if `GENERATE_AVX512_CODE` is defined, should we not be ready to go (and hence no more need to test the avx512 related macros)?

Finally got back to this. Bad news, with the above patch it still doesn't compile; it fails here:
```
  CCLD     ompi_info
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/binutils-bin/2.33.1/ld: ../../../ompi/.libs/libmpi.so: undefined reference to `llvm.x86.avx512.pmaxu.d'
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/binutils-bin/2.33.1/ld: ../../../ompi/.libs/libmpi.so: undefined reference to `llvm.x86.avx512.pmins.d'
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/binutils-bin/2.33.1/ld: ../../../ompi/.libs/libmpi.so: undefined reference to `llvm.x86.avx512.pmaxs.q'
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/binutils-bin/2.33.1/ld: ../../../ompi/.libs/libmpi.so: undefined reference to `llvm.x86.avx512.pmaxu.q'
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/binutils-bin/2.33.1/ld: ../../../ompi/.libs/libmpi.so: undefined reference to `llvm.x86.avx512.pminu.d'
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/binutils-bin/2.33.1/ld: ../../../ompi/.libs/libmpi.so: undefined reference to `llvm.x86.avx512.pminu.q'
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/binutils-bin/2.33.1/ld: ../../../ompi/.libs/libmpi.so: undefined reference to `llvm.x86.avx512.pmins.q'
```
 
I reduced it to an example with these 8 AVX512 functions:

```
#include <immintrin.h>

int main(int argc, char **argv)
{
    __m512i vecA = _mm512_loadu_si512((__m512 *)argv[0]);
    __m512i vecB = _mm512_loadu_si512((__m512 *)argv[1]);
    vecB = _mm512_max_epi64(vecA, vecB);
    vecB = _mm512_min_epi64(vecA, vecB);
    vecB = _mm512_max_epu64(vecA, vecB);
    vecB = _mm512_min_epu64(vecA, vecB);
    vecB = _mm512_max_epi32(vecA, vecB);
    vecB = _mm512_min_epi32(vecA, vecB);
    vecB = _mm512_max_epu32(vecA, vecB);
    vecB = _mm512_min_epu32(vecA, vecB);
    return (int)((char *)&vecB)[0];
}
```

```
$ nvc -tp skylake op_avx.c 
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/binutils-bin/2.33.1/ld: /tmp/nvclwCcHEoai04S.o: in function `_mm512_max_epu32':
/cvmfs/restricted.computecanada.ca/easybuild/software/2020/Core/nvhpc/22.1/Linux_x86_64/22.1/compilers/include/avx512fintrin.h:1111: undefined reference to `llvm.x86.avx512.pmaxu.d'
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/binutils-bin/2.33.1/ld: /tmp/nvclwCcHEoai04S.o: in function `_mm512_max_epi64':
/cvmfs/restricted.computecanada.ca/easybuild/software/2020/Core/nvhpc/22.1/Linux_x86_64/22.1/compilers/include/avx512fintrin.h:1133: undefined reference to `llvm.x86.avx512.pmaxs.q'
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/binutils-bin/2.33.1/ld: /tmp/nvclwCcHEoai04S.o: in function `_mm512_max_epu64':
/cvmfs/restricted.computecanada.ca/easybuild/software/2020/Core/nvhpc/22.1/Linux_x86_64/22.1/compilers/include/avx512fintrin.h:1155: undefined reference to `llvm.x86.avx512.pmaxu.q'
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/binutils-bin/2.33.1/ld: /tmp/nvclwCcHEoai04S.o: in function `_mm512_min_epi32':
/cvmfs/restricted.computecanada.ca/easybuild/software/2020/Core/nvhpc/22.1/Linux_x86_64/22.1/compilers/include/avx512fintrin.h:1324: undefined reference to `llvm.x86.avx512.pmins.d'
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/binutils-bin/2.33.1/ld: /tmp/nvclwCcHEoai04S.o: in function `_mm512_min_epu32':
/cvmfs/restricted.computecanada.ca/easybuild/software/2020/Core/nvhpc/22.1/Linux_x86_64/22.1/compilers/include/avx512fintrin.h:1346: undefined reference to `llvm.x86.avx512.pminu.d'
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/binutils-bin/2.33.1/ld: /tmp/nvclwCcHEoai04S.o: in function `_mm512_min_epi64':
/cvmfs/restricted.computecanada.ca/easybuild/software/2020/Core/nvhpc/22.1/Linux_x86_64/22.1/compilers/include/avx512fintrin.h:1368: undefined reference to `llvm.x86.avx512.pmins.q'
/cvmfs/soft.computecanada.ca/gentoo/2020/usr/x86_64-pc-linux-gnu/binutils-bin/2.33.1/ld: /tmp/nvclwCcHEoai04S.o: in function `_mm512_min_epu64':
/cvmfs/restricted.computecanada.ca/easybuild/software/2020/Core/nvhpc/22.1/Linux_x86_64/22.1/compilers/include/avx512fintrin.h:1390: undefined reference to `llvm.x86.avx512.pminu.q'
```

Bug reported to NVIDIA: https://forums.developer.nvidia.com/t/issue-with-some-avx512-intrinsics-min-max-ep-i-u-32-64/200868

@bosilca 
```console
mca/op/avx/.libs/libmca_op_avx.a(liblocal_ops_avx512_la-op_avx_functions.o):(.data+0x0): multiple definition of `ompi_op_avx_functions_avx2'
mca/op/avx/.libs/libmca_op_avx.a(liblocal_ops_avx2_la-op_avx_functions.o):(.data+0x0): first defined here
mca/op/avx/.libs/libmca_op_avx.a(liblocal_ops_avx512_la-op_avx_functions.o):(.data+0x1340): multiple definition of `ompi_op_avx_3buff_functions_avx2'
mca/op/avx/.libs/libmca_op_avx.a(liblocal_ops_avx2_la-op_avx_functions.o):(.data+0x1340): first defined here
make[2]: *** [Makefile:3291: libmpi.la] Error 1
```

--- a/ompi/mca/op/avx/configure.m4
+++ b/ompi/mca/op/avx/configure.m4
@@ -38,7 +38,13 @@ AC_DEFUN([MCA_ompi_op_avx_CONFIG],[
 
     OPAL_VAR_SCOPE_PUSH([op_avx_cflags_save])
 
-    AS_IF([test "$opal_cv_asm_arch" = "X86_64"],
+    case "${host}" in
+        x86_64-*x32|i?86-*|x86_64*|amd64*)
+            check_avx="yes";;
+        *)
+            check_avx="no";;
+    esac
+    AS_IF([test "$check_avx" = "yes"],
           [AC_LANG_PUSH([C])
 
            #
@@ -117,6 +123,28 @@ AC_DEFUN([MCA_ompi_op_avx_CONFIG],[
 #endif
     __m512i vA, vB;
     _mm512_mullo_epi64(vA, vB)
+                                      ]])],
+                             [AC_MSG_RESULT([yes])],
+                             [op_avx512_support=0
+                              MCA_BUILD_OP_AVX512_FLAGS=""
+                              AC_MSG_RESULT([no])])
+                         CFLAGS="$op_avx_cflags_save"
+                        ])
+                  #
+                  # Check for combination of AVX512F + AVX512VL
+                  #
+                  AS_IF([test $op_avx512_support -eq 1],
+                        [AC_MSG_CHECKING([if _mm_max_epi64 generates code that can be compiled])
+                         op_avx_cflags_save="$CFLAGS"
+                         CFLAGS="$CFLAGS_WITHOUT_OPTFLAGS -O0 $MCA_BUILD_OP_AVX512_FLAGS"
+                         AC_LINK_IFELSE(
+                             [AC_LANG_PROGRAM([[#include <immintrin.h>]],
+                                      [[
+#if !defined(__AVX512F__) || !defined(__AVX512VL__) || !defined(__AVX512BW__)
+#error "icc needs the -m flags to provide the AVX* detection macros"
+#endif
+    __m128i vA, vB;
+    _mm_max_epi64(vA, vB)
                                       ]])],
                              [AC_MSG_RESULT([yes])],
                              [op_avx512_support=0

--- a/ompi/mca/op/avx/op_avx_functions.c
+++ b/ompi/mca/op/avx/op_avx_functions.c
@@ -32,16 +32,18 @@
  * to a lesser support (AVX512 -> AVX2, AVX2 -> AVX, AVX -> error out).
  */
 #if defined(GENERATE_AVX512_CODE)
+#  define PREPEND _avx512
 #  if defined(__AVX512BW__) && defined(__AVX512F__) && defined(__AVX512VL__)
-#    define PREPEND _avx512
+/* all good */
 #  else
 #    undef GENERATE_AVX512_CODE
 #  endif  /* defined(__AVX512BW__) && defined(__AVX512F__) && defined(__AVX512VL__) */
 #endif  /* defined(GENERATE_AVX512_CODE) */
 
 #if !defined(PREPEND) && defined(GENERATE_AVX2_CODE)
+#  define PREPEND _avx2
 #  if defined(__AVX2__)
-#    define PREPEND _avx2
+/* all good */
 #  else
 #    undef GENERATE_AVX2_CODE
 #  endif  /* defined(__AVX2__) */

